File: 945-minimum-increment-to-make-array-unique/945-minimum-increment-to-make-array-unique.go
func minIncrementForUnique(nums []int) int {
    sort.Ints(nums)
    n := len(nums)
    duplicates := 0
    ans := 0
    for i:=1;i<n;i+=1 {
        if nums[i] == nums[i-1] {
            duplicates++
            ans -= nums[i]
        }else{
            option1 := nums[i] - nums[i-1] - 1
            option2 := duplicates
            var numbers int
            if option1 < option2 {
                numbers = option1
            }else {
                numbers = option2
            }
            add := numbers * nums[i-1] + (numbers*(numbers+1))/2
            duplicates -= numbers
            ans += add
        }
    }
    if duplicates > 0 {
        ans += duplicates * nums[n-1] + (duplicates*(duplicates+1))/2
        duplicates = 0
    }
    return ans
}

File: 945-minimum-increment-to-make-array-unique/945-minimum-increment-to-make-array-unique.cpp
class Solution {
public:
    int minIncrementForUnique(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        int n = nums.size(),ans = 0,duplicates = 0;
        for(int i=1;i<n;i++){
            if(nums[i] == nums[i-1]){
                duplicates++;
                ans -= nums[i];
            }else{
                int numsPlaced = min(nums[i]-nums[i-1]-1,duplicates);
                int add = numsPlaced * nums[i-1] + (numsPlaced*(numsPlaced+1))/2;
                ans += add;
                duplicates -= numsPlaced;
            }
            //cout<<"ans: "<<ans<<"\n";
        }
        if(duplicates > 0){
            int add = (duplicates*(duplicates+1))/2 + duplicates*nums.back();
            ans += add;
            duplicates = 0;
            //cout<<"ans here: "<<ans<<"\n";
        }
        return ans;
    }
};